views/components/Ad.jsx

```diff
@@ -51,14 +53,52 @@ class Ad extends React.Component {
   }
 
   componentDidMount () {
-    var ctx = this;
-
     this.getAd().then((ad) => {
+      this.adObject = new models.Link(ad).toJSON();
       return this.setState({
         loaded: true,
-        ad: new models.Link(ad).toJSON(),
+        ad: this.adObject,
       });
     }, () => {});
+
+    window.addEventListener('scroll', this._onScroll);
+    window.addEventListener('resize', this._onScroll);
+    this._hasListeners = true;
+    this._onScroll();
+  }
+
+  componentWillUnmount() {
+    this._removeListeners();
+  }
+
+  _removeListeners() {
+    if (this._hasListeners) {
+      window.removeEventListener('scroll', this._onScroll);
+      window.removeEventListener('resize', this._onScroll);
+      this._hasListeners = false;
+    }
+  }
+
+  _onScroll() {
+    var node = React.findDOMNode(this);
+    var winHeight = window.innerHeight;
+    var rect = node.getBoundingClientRect();
+    var top = rect.top;
+    var height = rect.height;
+    var bottom = top + rect.height;
+    var middle = (top + bottom) / 2;
+    var middleIsAboveBottom = middle < winHeight;
+    var middleIsBelowTop = bottom > 45 + height / 2;
+
+    if(middleIsAboveBottom && middleIsBelowTop) {
+      var srcs=['imp_pixel', 'adserver_imp_pixel'];
+      for (var i = 0, iLen = srcs.length; i < iLen; i++) {
+        var pixel = new Image();
+        pixel.src = this.adObject[srcs[i]];
```

Comment:
I don't believe this breaks when no ads are displayed – I can't see any errors, at any rate.

This component is only invoked by `mutators.jsx` when an ad is available. If there is no ad, the component is never mounted and none of this code should run.

Or am I missing something?